resource.go

```diff
@@ -0,0 +1,109 @@
+	monitoredrespb "google.golang.org/genproto/googleapis/api/monitoredres"
+)
+
+type resourceMap struct {
+	// Mapping from the input resource type to the monitored resource type in Stackdriver.
+	srcType, dstType string
+	// Mapping to Stackdriver monitored resource label to an OpenCensus resource label.
+	labels map[string]string
+}
+
+// Resource labels that are generally internal to the exporter.
+// Consider exposing these labels and a type identifier in the future to allow
+// for customization.
+const (
+	stackdriverLocation             = "contrib.opencensus.io/exporter/stackdriver/location"
+	stackdriverProjectID            = "contrib.opencensus.io/exporter/stackdriver/project_id"
+	stackdriverGenericTaskNamespace = "contrib.opencensus.io/exporter/stackdriver/generic_task/namespace"
+	stackdriverGenericTaskJob       = "contrib.opencensus.io/exporter/stackdriver/generic_task/job"
+	stackdriverGenericTaskID        = "contrib.opencensus.io/exporter/stackdriver/generic_task/task_id"
+)
+
+// Mappings for the well-known OpenCensus resources to applicable Stackdriver resources.
+var resourceMappings = []resourceMap{
+	{
+		srcType: resourcekeys.K8STypeContainer,
+		dstType: "k8s_container",
+		labels: map[string]string{
+			"project_id":     stackdriverProjectID,
+			"location":       stackdriverLocation,
+			"cluster_name":   resourcekeys.K8SKeyClusterName,
+			"namespace_name": resourcekeys.K8SKeyNamespaceName,
+			"pod_name":       resourcekeys.K8SKeyPodName,
+			"container_name": resourcekeys.K8SKeyContainerName,
+		},
+	},
+	{
+		srcType: resourcekeys.GCPTypeGCEInstance,
+		dstType: "gce_instance",
+		labels: map[string]string{
+			"project_id":  resourcekeys.GCPKeyGCEProjectID,
+			"instance_id": resourcekeys.GCPKeyGCEInstanceID,
+			"zone":        resourcekeys.GCPKeyGCEZone,
+		},
+	},
+	{
+		srcType: resourcekeys.AWSTypeEC2Instance,
+		dstType: "aws_ec2_instance",
+		labels: map[string]string{
+			"project_id":  stackdriverProjectID,
+			"instance_id": resourcekeys.AWSKeyEC2InstanceID,
+			"region":      resourcekeys.AWSKeyEC2Region,
+			"aws_account": resourcekeys.AWSKeyEC2AccountID,
+		},
+	},
+	// Fallback to generic task resource.
+	{
+		srcType: "",
+		dstType: "generic_task",
+		labels: map[string]string{
+			"project_id": stackdriverProjectID,
+			"location":   stackdriverLocation,
+			"namespace":  stackdriverGenericTaskNamespace,
+			"job":        stackdriverGenericTaskJob,
+			"task_id":    stackdriverGenericTaskID,
+		},
+	},
+}
+
+func DefaultMapResource(res *resource.Resource) *monitoredrespb.MonitoredResource {
+Outer:
+	for _, rm := range resourceMappings {
+		if res.Type != rm.srcType {
+			continue
+		}
+		result := &monitoredrespb.MonitoredResource{
+			Type:   rm.dstType,
+			Labels: make(map[string]string, len(rm.labels)),
+		}
+		for dst, src := range rm.labels {
+			if v, ok := res.Labels[src]; ok {
+				result.Labels[dst] = v
+			} else {
+				// A required label wasn't filled at all. Try subsequent mappings.
```

Comment:
This implies that we can have multiple mappings with the same srcType. That doesn't seem needed for the types we support, so should we just return right away?

Reply:
@jkohen If we returned immediately, we wouldn't fall through to the generic variant if no type is set. We could of course just do a special case at the end. Both options seemed to work.

@rghetia If the user configured a type, all labels must be provided to the Stackdriver API. If they are not, there's mis-configuration which should be resolved by the user. If the user didn't specify a type, we want to fallback to the generic task resource type.
Trying to do greedy matching like 2/3 suggest can't be processed by the Stackdriver API, thus there's no point in trying that.

Reply:
@fabxc Sorry if I'm missing something obvious. Is the "generic variant" `generic_task`? I still don't understand how it gets matched, because `if (res.Type != rm.srcType) continue` is equivalent to `if (res.Type != "") continue`, and I don't see why that would be true. I think there is no test case covering that either.